experimental/ragged_inference/triton_v2_ragged_qk_dotprod.py

```diff
@@ -0,0 +1,348 @@
+from dataclasses import dataclass
+from typing import List, Optional
+
+import torch
+import triton
+import triton.language as tl
+from ragged_inference.garbage_pad_ragged_acts import RaggedActivations
+from triton.ops.matmul_perf_model import estimate_matmul_time, prune_num_stages
+
+
+def init_to_zero(name):
+    return lambda nargs: nargs[name].zero_()
+
+
+# TODO: tune these
+BLOCK_Q = 16
+BLOCK_K = 128
+BLOCK_D = 32
+
+
+def get_fast_dev_configs():
+    return [
+        triton.Config(
+            {"BLOCK_Q": BLOCK_Q, "BLOCK_K": BLOCK_K, "BLOCK_D": BLOCK_D},
+            num_stages=5,
+            num_warps=2,
+        )
+    ]
+
+
+@triton.autotune(
+    # configs=get_all_configs(),
+    configs=get_fast_dev_configs(),
+    key=["max_n_ctx_q_across_seqs", "max_n_ctx_k_across_seqs", "d_head"],
+    prune_configs_by={
+        "prune_num_stages_by": prune_num_stages,
+        "perf_model": estimate_matmul_time,
+        "top_k": 10,
+    },
+)
+@triton.jit
+def _qk_dotprod_kernel(
+    # Pointers to our tensors
+    q_ptr,
+    k_ptr,
+    scores_ptr,  # Rectangular output tensor
+    # Pointers to lookup tables (sometimes referred to as a "lut")
+    pid_to_in_q_token_offset_ptr,
+    pid_to_in_k_token_offset_ptr,
+    pid_to_out_q_block_ptr,
+    pid_to_out_k_block_ptr,
+    pid_to_out_seq_idx_ptr,
+    # Integers
+    max_n_ctx_q_across_seqs,
+    max_n_ctx_k_across_seqs,
+    d_head,
+    stride_ctx_q,
+    stride_ctx_k,
+    stride_out_q,
+    stride_out_k,
+    stride_out_seq,
+    total_ctx_q_across_all_seqs,
+    total_ctx_k_across_all_seqs,
+    # These get populated from the triton.Config
+    BLOCK_Q: tl.constexpr,
+    BLOCK_K: tl.constexpr,
+    BLOCK_D: tl.constexpr,
+):
+    """
+    Adapted from https://github.com/openai/triton/blob/v2.0/python/triton/ops/matmul.py
+    """
+
+    # matrix multiplication
+    pid = tl.program_id(0)
+
+    out_q_block = tl.load(pid_to_out_q_block_ptr + pid)
+    out_k_block = tl.load(pid_to_out_k_block_ptr + pid)
+    out_seq_idx = tl.load(pid_to_out_seq_idx_ptr + pid)
+    in_q_token_offset = tl.load(pid_to_in_q_token_offset_ptr + pid)
+    in_k_token_offset = tl.load(pid_to_in_k_token_offset_ptr + pid)
+
+    # Define indices ranges, we follow the triton convention of prefixing
+    # with "r" to denote a range like "rq" is the range for queries below
+    rq = in_q_token_offset + tl.arange(0, BLOCK_Q)
```

Comment:
nice, "all" (well, a lot of) the magic is there, looks like nothing but super well done I think